CUMULUS-3076: Add Workflow Message Granule Writes docs (#3390)
nasa/cumulus: docs/workflows/message_granule_writes.md

---
id: message_granule_writes
title: Workflow Message Granule Writes
hide_title: false
---

## Overview

When an [AWS Step Function Event](https://docs.aws.amazon.com/step-functions/latest/dg/cw-events.html) occurs for a [Cumulus workflow](https://nasa.github.io/cumulus/docs/next/workflows/) *or* a write is attempted via the [sf-sqs-report task](https://github.com/nasa/cumulus/tree/master/tasks/sf-sqs-report) a message is dispatched to the `sfEventSqsToDbRecordsInputQueue` for processing.

> The `sfEventSqsToDbRecords` is then triggered in batches of 10 messages each, which correspond to lambda invocations or workflow events, and the corresponding execution/PDR is attempted to write, then the granule records associated with the message are also attempted to be written.

Reviewer: ```suggestion
Messages on the `sfEventSqsToDbRecordsInputQueue` (which correspond to lambda invocations or workflow events) are processed in batches of 10 and the `sfEventSqsToDbRecords` Lambda is triggered for each. The corresponding execution/PDR is attempted to write, then the granule records associated with the message are also attempted to be written.
```

> For each granule in the batch of granules one of the following occurs:

Reviewer: ```suggestion
For each granule in the batch of granules **one of the following** occurs:
```

- The granule is written successfully.
- The granule write is dropped, due to asynchronous write constraints.

> - The lambda fails to write the granule in an unexpected way (e.g. lambda failure, AWS outage, etc).   In this case, the granule will become visible again after the `sfEventSqsToDbRecordsInputQueue` visibility timeout configuration (currently set as a function of the rds_connection_timing_configuration terraform variable:

Reviewer: ```suggestion
- The lambda fails to write the granule in an unexpected way (e.g. lambda failure, AWS outage, etc).   In this case, the granule will become visible again after the `sfEventSqsToDbRecordsInputQueue` visibility timeout (currently set as a function of the rds_connection_timing_configuration terraform variable:
```

```terraform
var.rds_connection_timing_configuration.acquireTimeoutMillis / 1000) + 60
```

> - The granule fails to write due to a schema violation, database connection issue or other expected/trapped error.    The message is immediately written to the [Dead Letter Archive](https://nasa.github.io/cumulus/docs/features/dead_letter_archive/) for User manual intervention/investigation.

Reviewer: ```suggestion
- The granule fails to write due to a schema violation, database connection issue or other expected/caught error.    The message is immediately written to the [Dead Letter Archive](https://nasa.github.io/cumulus/docs/features/dead_letter_archive/) for manual intervention/investigation.
```

### Caveats

- All non-bulk [Cumulus API](https://nasa.github.io/cumulus-api/) granule operations are *not* constrained by this logic and do not utilize the SQS update queue.  They are instead invoked synchronously and follow expected RESTful logic without any asynchronous write constraints *or* default message values.
- This information is correct as of release v16 of Cumulus Core.   Please review the [CHANGELOG](https://github.com/nasa/cumulus/blob/master/CHANGELOG.md) and migration instructions for updated features/changes/bugfixes.

## Granule Write Constraints

For each granule to be written, the following constraints apply:

> - `granuleId` must be unique.

Reviewer: What actually happens if it's not unique? The write will fail because ElasticSearch indexing fails?

  Granule write will not be allowed if `granuleId` already exists in the database for another collection, granules in this state will be rejected to write and wind up in the [Dead Letter Archive](https://nasa.github.io/cumulus/docs/features/dead_letter_archive/)

> - Message granule must match the API schema.

Reviewer: ```suggestion
- Message granule must match the API Granule schema.
```
link maybe?

Author: Good suggestion: 39db55f39c

>   If not the write will be rejected, the granule status will be updated to failed, and the message will wind up in the [Dead Letter Archive](https://nasa.github.io/cumulus/docs/features/dead_letter_archive/)

Reviewer: ```suggestion
  If not the write will be rejected, the granule status will be updated to `failed`, and the message will wind up in the [Dead Letter Archive](https://nasa.github.io/cumulus/docs/features/dead_letter_archive/)
```

> - If the granule is being updated to a running/queued state:

Reviewer: ```suggestion
- If the granule is being updated to a `running`/`queued` status:
```

>   - Only `status`, `timestamp`, `updated_at` and `created_at` are updated.   All other values are retained as-they currently exist in the database.

Reviewer: ```suggestion
  - Only `status`, `timestamp`, `updated_at` and `created_at` are updated.   All other values are retained as they currently exist in the database.
```
